File: solutions/400_nth_digit.rb
# Find the nth digit of the infinite integer sequence 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ...
#
# Note:
# n is positive and will fit within the range of a 32-bit signed integer (n < 231).
#
# Input:
# 11
#
# Output:
# 0
#
# Explanation:
# The 11th digit of the sequence 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ... is a 0, which is part of the number 10.

# @param {Integer} n
# @return {Integer}
def find_nth_digit(n)
  s = [1, 1]
  while n > s[1]
    s[0] += 1
    s[1] += s[0].to_s.length
  end
  s[0].to_s[n - s[1] - 1]
end

def find_nth_digit2(n)
  return n if n < 10
  cache = [[1, 9, 9], [2, 99, 189], [3, 999, 2889], [4, 9999, 38_889], [5, 99_999, 488_889], [6, 999_999, 5_888_889],
           [7, 9_999_999, 68_888_889], [8, 99_999_999, 788_888_889], [9, 999_999_999, 8_888_888_889]]
  i = 0
  i += 1 while cache[i].last < n
  return cache[i][1].to_s[-1].to_i if cache[i].last == n

  quo = (n - cache[i - 1][-1]) / cache[i][0]
  rem = (n - cache[i - 1][-1]) % cache[i][0]
  return (quo + cache[i - 1][1]).to_s[-1].to_i if rem.zero?
  (quo + cache[i - 1][1] + 1).to_s[rem - 1].to_i
end

(1..10).each do |n|
  p find_nth_digit(n)
end
p 2**31 - 1

10.times.each do |_i|
  n = Random.rand(2**10 - 1)
  n = 38_889
  p [n, find_nth_digit(n), find_nth_digit2(n)]
end

File: python_solutions/400.nth-digit.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left as bl, bisect_right as br
from functools import reduce
import string
true = True
false = False
MIN, MAX = -0x3f3f3f3f, 0x3f3f3f3f
#
# @lc app=leetcode id=400 lang=python3
#
# [400] Nth Digit
#
# https://leetcode.com/problems/nth-digit/description/
#
# algorithms
# Medium (30.86%)
# Total Accepted:    52.9K
# Total Submissions: 171.3K
# Testcase Example:  '3'
#
# Find the n^th digit of the infinite integer sequence 1, 2, 3, 4, 5, 6, 7, 8,
# 9, 10, 11, ...
#
# Note:
# n is positive and will fit within the range of a 32-bit signed integer (n <
# 2^31).
#
#
# Example 1:
#
# Input:
# 3
#
# Output:
# 3
#
#
#
# Example 2:
#
# Input:
# 11
#
# Output:
# 0
#
# Explanation:
# The 11th digit of the sequence 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, ... is a 0,
# which is part of the number 10.
#
#
#


class Solution:
    def findNthDigit(self, n: int) -> int:
        if n < 10:
            return n
        base, bit, cc = 1, 1, 0
        while true:
            cc += (base * 9) * bit
            if cc == n:
                return 9
            elif cc > n:
                break
            base *= 10
            bit += 1
        cc -= (base * 9) * bit
        q, r = divmod(n - cc, bit)

        q += pow(10, bit - 1) - 1
        # print(cc, base, bit, q, r)
        if r == 0:
            return int(str(q)[-1])
        else:
            return int(str(q + 1)[r - 1])


sol = Solution()
n = 1000
print(sol.findNthDigit(n))